File: algorithms/tokenBucket.py
import time

class TokenBucket:
    def __init__(self, capacity, refill_rate_per_second):
        self.capacity = capacity
        self.refill_rate_per_second = refill_rate_per_second
        self.tokens = capacity
        self.last_refill_timestamp = time.time()

    def allow_request(self):
        self.refill()
        if self.tokens >= 1:
            self.tokens -= 1
            return True
        return False

    def refill(self):
        now = time.time()
        elapsed = now - self.last_refill_timestamp
        tokens_to_add = int(elapsed * self.refill_rate_per_second)

        if tokens_to_add > 0:
            self.tokens = min(self.capacity, self.tokens + tokens_to_add)
            self.last_refill_timestamp = now

if __name__ == "__main__":
    bucket = TokenBucket(5, 1)  # capacity of 5 tokens, refills 1 token per second

    for i in range(10):
        allowed = bucket.allow_request()
        print(f"Request {i}: {'Allowed' if allowed else 'Blocked'}")
        time.sleep(0.5)  # simulate 0.5 second between requests

File: algorithms/TokenBucket.java
import java.util.concurrent.atomic.AtomicInteger;

public class TokenBucket {
    private final int capacity;
    private final int refillRatePerSecond;
    private int tokens;
    private long lastRefillTimestamp;

    public TokenBucket(int capacity, int refillRatePerSecond) {
        this.capacity = capacity;
        this.refillRatePerSecond = refillRatePerSecond;
        this.tokens = capacity;
        this.lastRefillTimestamp = System.nanoTime();
    }

    public synchronized boolean allowRequest() {
        refill();

        if (tokens > 0) {
            tokens--;
            return true;
        }

        return false;
    }

    private void refill() {
        long now = System.nanoTime();
        long elapsedTime = now - lastRefillTimestamp;
        int tokensToAdd = (int) ((elapsedTime / 1_000_000_000.0) * refillRatePerSecond);

        if (tokensToAdd > 0) {
            tokens = Math.min(capacity, tokens + tokensToAdd);
            lastRefillTimestamp = now;
        }
    }

    public static void main(String[] args) throws InterruptedException {
        TokenBucket bucket = new TokenBucket(5, 1); // 5 capacity, 1 token per second

        for (int i = 0; i < 10; i++) {
            System.out.println("Request " + i + ": " + (bucket.allowRequest() ? "Allowed" : "Blocked"));
            Thread.sleep(500); // simulate time delay
        }
    }
}

